kernel/src/engine/parquet_stats_skipping.rs

```diff
@@ -0,0 +1,1267 @@
+    /// "defaults" to the non-dominant value (and we can actually just ignore non-dominant inputs).
+    ///
+    /// If the filter is inverted, use de Morgan's laws to push the inversion down into the inputs
+    /// (e.g. `NOT(AND(a, b))` becomes `OR(NOT(a), NOT(b))`).
+    fn apply_variadic(
+        &self,
+        op: VariadicOperator,
+        exprs: &[Expression],
+        inverted: bool,
+    ) -> Option<bool> {
+        // With AND (OR), any FALSE (TRUE) input forces FALSE (TRUE) output.  If there was no
+        // dominating input, then any NULL input forces NULL output.  Otherwise, return the
+        // non-dominant value. Inverting the operation also inverts the dominant value.
+        let dominator = match op {
+            VariadicOperator::And => inverted,
+            VariadicOperator::Or => !inverted,
+        };
+
+        // Evaluate the input expressions, inverting each as needed and tracking whether we've seen
+        // any NULL result. Stop immediately (short circuit) if we see a dominant value.
+        let result = exprs.iter().try_fold(false, |found_null, expr| {
+            match self.apply_expr(expr, inverted) {
+                Some(v) if v == dominator => None, // (1) short circuit, dominant found
+                Some(_) => Some(found_null),
+                None => Some(true), // (2) null found (but keep looking for a dominant value)
+            }
+        });
+
+        match result {
+            None => Some(dominator), // (1) short circuit, dominant found
+            Some(false) => Some(!dominator),
+            Some(true) => None, // (2) null found, dominant not found
+        }
+    }
+
+    /// Evaluates binary comparisons. Any NULL input produces a NULL output. If `inverted`, the
+    /// opposite operation is performed, e.g. `<` evaluates as if `>=` had been requested instead.
+    fn apply_binary(
+        &self,
+        op: BinaryOperator,
+        left: &Expression,
+        right: &Expression,
+        inverted: bool,
+    ) -> Option<bool> {
+        use BinaryOperator::*;
+        use Expression::{Column, Literal};
+
+        // Min/Max stats don't allow us to push inversion down into the comparison. Instead, we
+        // invert the comparison itself when needed and compute normally after that.
+        let op = match inverted {
+            true => op.invert()?,
+            false => op,
+        };
+
+        // NOTE: We rely on the literal values to provide logical type hints. That means we cannot
+        // perform column-column comparisons, because we cannot infer the logical type to use.
+        let (op, col, val) = match (left, right) {
+            (Column(col), Literal(val)) => (op, col, val),
+            (Literal(val), Column(col)) => (op.commute()?, col, val),
+            (Literal(a), Literal(b)) => return Self::apply_binary_scalars(op, a, b),
+            _ => None?, // unsupported combination of operands
```

Comment:
minor point, but do you find `None?` easier to read than `return None` is?

Reply:
I personally find `None?` to be a bit more confusing to read, maybe I'm reading too much into the natural language of it, but it just seems like you aren't sure if it's none or not. I know that's not what it literally means as it's shorthand, but my point is I do a double take on what it means the first time I see it.

Reply:
I can switch it back. Agree it's a bit too "clever"